<?xml version="1.0"?>
<launch>
  <!-- Choose the grinding mode for the Bezier library -->
  <arg name="surfacing_mode" default="false"/>
  <param name="surfacing_mode_param" value="$(arg surfacing_mode)" />
  <!-- Specify CAD mesh for Bezier -->
  <arg name="mesh_cad" />
  <param name="mesh_cad_param" value="$(arg mesh_cad)" />
  <!-- Specify defect mesh for Bezier (only required if NOT using surfacing mode) -->
  <arg name="mesh_defect" default="" />
  <param name="mesh_defect_param" value="$(arg mesh_defect)" />

  <!-- The planning and execution components of MoveIt! configured to run -->
  <!-- using the ROS-Industrial interface. -->

  <!-- Non-standard joint names:
       - Create a file [robot_moveit_config]/config/joint_names.yaml
           controller_joint_names: [joint_1, joint_2, ... joint_N]
       - Update with joint names for your robot (in order expected by rbt controller)
       - and uncomment the following line: -->
  <rosparam command="load" file="$(find fanuc_m20ia_support)/config/joint_names_m20ia.yaml" />

  <!-- the "sim" argument controls whether we connect to a Simulated or Real robot -->
  <!--  - if sim=false, a robot_ip argument is required -->
  <arg name="sim" default="true" />
  <arg name="robot_ip" unless="$(arg sim)" />
  <arg name="use_bswap" unless="$(arg sim)" default="true" />

  <!-- load the robot_description parameter before launching ROS-I nodes -->
  <include file="$(find fanuc_m20ia_moveit_config)/launch/planning_context.launch" >
    <arg name="load_robot_description" value="true" />
  </include>

  <!-- run the robot simulator and action interface nodes -->
  <group if="$(arg sim)">
    <include file="$(find industrial_robot_simulator)/launch/robot_interface_simulator.launch" />
  </group>

  <!-- run the "real robot" interface nodes -->
  <!--   - this typically includes: robot_state, motion_interface, and joint_trajectory_action nodes -->
  <!--   - replace these calls with appropriate robot-specific calls or launch files -->
  <group unless="$(arg sim)">
    <include file="$(find fanuc_m20ia_support)/launch/robot_interface_streaming_m20ia.launch" >
      <arg name="robot_ip" value="$(arg robot_ip)"/>
      <arg name="use_bswap"  value="$(arg use_bswap)" />
    </include>
  </group>

  <!-- publish the robot state (tf transforms) -->
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />

  <include file="$(find fanuc_m20ia_moveit_config)/launch/move_group.launch">
    <arg name="publish_monitored_planning_scene" value="true" />
  </include>

  <node name="bezier_application_surfacing" pkg="bezier_application" type="bezier_application_surfacing" output="screen" required="true" />

  <include file="$(find fanuc_m20ia_support)/launch/load_m20ia.launch" />
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find bezier_application)/config/robot_state_visualize.rviz" required="true" />
</launch>

